### ISSUE TITLE
Boolean and None defaults

### ISSUE THREAD

**Issue by brynpickering** [calliope-project/calliope#260]
## Problem description

There are a number of defaults which are either boolean (True/False) or None. 

### Bool
Usually this doesn't cause issues. But, the defaults make their way back into the xarray dataset of the model data if you access the backend (`model.backend.access_model_inputs()` or `model.backend.rerun()`).
But, when saving to file, xarray will complain (and raise an exception) if it finds a boolean value as the first entry of a dataarray. If it doesn't find the boolean value here, NETCDF4 will serialize the data to integers, which remain as integers when loading the data back from file. Both are sub-optimal.

### None

Some defaults that are None have the ability to make it into Pyomo, namely those which are associated with `_equals` group constraints. E.g. applying `cost_max_equals` to only one cost class, when the model has multiple cost classes, will lead to NaN being in Pyomo for all the undefined constraints. Pyomo will now raise an exception unless the constraint actively ignores the occurrence of NaNs (which should probably be happening).

Both issues are caused by our choice of defaults. If a constraint is effectively switched off, how should a user / the defaults define this? Some constraints default to False, some to None...

Some constraints are boolean, so False makes sense as the default, but this will make xarray unhappy. These should probably be set to 0/1, but should they be set as such in the defaults, and we purge all mention of 'true'/'false'? The latter is easier for a user to understand, but if we let users define 'true'/'false' then, at some point, we need to convert everything to 0/1. Doing that internally may confuse a user who then looks at the inputs in the xarray dataset.

In conclusion:  

1. We should be uniform in the choice of type to define any defaults that supposed to default to not existing. If 'None', resulting 'None'/NaN entries in the xarray dataset need to be handled properly.

2. For boolean constraints, we need to decide between enforcing only the use of 0/1 in the YAML definition, or do something in preprocessing to undertake this conversion.

## Calliope version

0.6.5-dev (634ffd4c0db5e60ea0592270bc02a7f514442341)

**Comment by timtroendle** [comment]
Stumbled upon this today, because this (rather usual I suppose) use case crashes:

```Python
import calliope
model = calliope.examples.national_scale() # or any other model
# optionally change model params here
new_results = model.backend.rerun()
new_results.to_netcdf("test.nc")
```

**Comment by brynpickering** [comment]
Another use case which causes model failure due to this issue is defining a group constraint with `carrier_prod_share_min`/`max`/`equals`. Any instance where a defined technology isn't associated with one of the defined carriers, an exception occurs. E.g.:

```yaml
group_constraints:
    foo:
        techs: [A, B, C]
        carrier_prod_share_min:
            electricity: 0.1  # this isn't one of the carriers associated with C -> Pyomo raises an exception
```

**Comment by sjpfenninger** [comment]
Closing as addressed in #326
